src/requests/errors.rs

```diff
@@ -0,0 +1,270 @@
+use std::{error::Error, fmt::Debug, str::from_utf8};
+
+use tonic::codegen::http;
+
+use crate::auth::AuthError;
+
+#[derive(Debug, Clone)]
+pub enum MomentoErrorCode {
+    /// Invalid argument passed to Momento client
+    InvalidArgumentError,
+    /// Service returned an unknown response
+    UnknownServiceError,
+    /// Resource with specified name already exists
+    AlreadyExistsError,
+    /// Cache with specified name doesn't exist
+    NotFoundError,
+    /// An unexpected error occurred while trying to fulfill the request
+    InternalServerError,
+    /// Insufficient permissions to perform operation
+    PermissionError,
+    /// Invalid authentication credentials to connect to service
+    AuthenticationError,
+    /// Request was cancelled by the server
+    CancelledError,
+    /// Request rate, bandwidth, or object size exceeded the limits for the account
+    LimitExceededError,
+    /// Request was invalid
+    BadRequestError,
+    /// Client's configured timeout was exceeded
+    TimeoutError,
+    /// Server was unable to handle the request
+    ServerUnavailable,
+    /// A client resource (most likely memory) was exhausted
+    ClientResourceExhausted,
+    /// System is not in a state required for the operation's execution
+    FailedPreconditionError,
+    /// Unknown error has occurred
+    UnknownError,
+    /// Cache request responded with a Miss
+    Miss,
+    /// Type error
+    TypeError,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("{details}")]
+pub struct MomentoGrpcErrorDetails {
+    pub code: tonic::Code,
+    pub details: String,
+    pub message: String,
+    pub metadata: tonic::metadata::MetadataMap,
+}
+
+impl From<tonic::Status> for MomentoGrpcErrorDetails {
+    fn from(status: tonic::Status) -> Self {
+        MomentoGrpcErrorDetails {
+            code: status.code(),
+            details: from_utf8(status.details()).unwrap_or_default().into(),
+            message: status.message().into(),
+            metadata: status.metadata().clone(),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("{message}")]
+pub struct MomentoError {
+    pub message: String,
+    pub error_code: MomentoErrorCode,
+    #[source]
+    pub inner_error: Option<ErrorSource>,
+    pub details: Option<MomentoGrpcErrorDetails>,
+}
```

Comment:
Switched MomentoError from an enum to a struct with publicly accessible methods. Example of the UX:

```
let result = cache_client.create_cache(CACHE_NAME.to_string()).await;
    match result {
    Ok(_) => println!("Successfully created cache '{}'", CACHE_NAME),
    Err(e) => println!("Error code: {:?} | Message: {}", e.error_code, e.message)
}
```

Reply:
I think I like that - can you still do an `if let` on the `Err` branch like you did in your code in your earlier comment above?

Also I just noticed the `AlreadyExists` stuff above, which we moved to be a non-error case in the CreateCache response in the other SDKs, probably worth getting that in while we are messing with this.

Reply:
Didn't remove the AlreadyExists error completely since that can still be an error code received from GRPC, but the CreateCache response is now an enum and allows the possibility of checking if the Ok response was Created or AlreadyExists:

```
let result = cache_client.create_cache(CACHE_NAME.to_string()).await;
if let Err(e) = result {
    println!("Error: {}", e);
} else if let Ok(ok) = result {
    match ok {
        CreateCache::Created => println!("Successfully created cache '{}'", CACHE_NAME),
        CreateCache::AlreadyExists => println!("Cache '{}' already exists", CACHE_NAME),
    }
}
```

Reply:
> Makes sense, will remove `AlreadyExists` from the errors list.
> 
> Using `if let` could look something like this:
> 
> ```
> let result = cache_client.create_cache(CACHE_NAME.to_string()).await;
> if let Err(e) = result {
>     println!("Error: {}", e);
> } else {
>     println!("Successfully created cache '{}'", CACHE_NAME);
> }
> ```

Cool, thanks. I'm also interested in seeing if there's a way to do an if-let on the specific error type (maybe just the error code now), like you had done in your first example.

i.e., in your first example, I really liked how you could write conditions for the specific error types if you cared about them, but it seemed maybe hard to interact with errors generically. Now I think you've solved the "generically" case, and I want to make sure it's still also possible to drill in on the specific types if users want to.

Reply:
Yeah seems possible to use `if let` on the error codes instead:

```
let result = cache_client.create_cache(CACHE_NAME.to_string()).await;
if let Err(e) = result {
    if let MomentoErrorCode::InternalServerError = e.error_code {
        println!("InternalServerError: {}", e);
        return Ok(());
    } else {
        println!("Error: {}", e);
    }
} else if let Ok(ok) = result {
    match ok {
        CreateCache::Created => println!("Successfully created cache '{}'", CACHE_NAME),
        CreateCache::AlreadyExists => println!("Cache '{}' already exists", CACHE_NAME),
    }
}
```

Reply:
i pulled this code down and played with it, and you can match on e.error_code too, so overall I think I am liking this pattern a lot better. thanks!